atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTracker.java

```diff
@@ -0,0 +1,132 @@
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableKvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.ImmutableTransactionReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.KvsCallReadInfo;
+import com.palantir.atlasdb.transaction.api.expectations.TransactionReadInfo;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public final class KeyValueServiceDataTracker {
+    private final AtomicLongMap<TableReference> bytesReadByTable = AtomicLongMap.create();
+    private final LongAdder bytesReadOverall = new LongAdder();
+    private final AtomicLongMap<TableReference> kvsCallByTable = AtomicLongMap.create();
+    private final LongAdder kvsCallsOverall = new LongAdder();
+    private final ConcurrentMap<TableReference, KvsCallReadInfo> maximumBytesKvsCallInfoByTable =
+            new ConcurrentHashMap<>();
+    private final AtomicReference<Optional<KvsCallReadInfo>> maximumBytesKvsCallInfoOverall =
+            new AtomicReference<>(Optional.empty());
+
+    public TransactionReadInfo getReadInfo() {
+        return ImmutableTransactionReadInfo.builder()
+                .bytesRead(bytesReadOverall.longValue())
+                .kvsCalls(kvsCallsOverall.longValue())
+                .maximumBytesKvsCallInfo(maximumBytesKvsCallInfoOverall.get())
+                .build();
+    }
+
+    /**
+     * This is un-synchronized as it is called after task completion and the abort/commit stage.
+     * For returned statistics, no consistency guarantees are provided if the user interacts with the transaction
+     * outside the context of {@link com.palantir.atlasdb.transaction.api.TransactionTask} or after commit/abort
+     * stages (e.g. this can happen with futures/iterators).
+     */
+    public ImmutableMap<TableReference, TransactionReadInfo> getReadInfoByTable() {
+        Set<TableReference> tableRefs =
+                ImmutableSet.copyOf(kvsCallByTable.asMap().keySet());
+
+        return tableRefs.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Function.identity(), tableRef -> ImmutableTransactionReadInfo.builder()
+                                .bytesRead(bytesReadByTable.get(tableRef))
+                                .kvsCalls(kvsCallByTable.get(tableRef))
+                                .maximumBytesKvsCallInfo(
+                                        Optional.ofNullable(maximumBytesKvsCallInfoByTable.get(tableRef)))
+                                .build()));
+    }
+
+    /**
+     * Tracks an effectively completed kvs read method call for some {@link TableReference}.
```

Comment:
Perhaps - `Tracks a fully resolved (i.e., non-lazy) Atlas KVS read call for a given {@link TableReference}. For lazy calls, see {@link recordCallForTable}`